content/en/docs/faq/timing-operations.md

```diff
@@ -0,0 +1,247 @@
+
+The following sample program implements a basic instrumentation probe in terms
+of `smithy-go` middleware for SQS clients and demonstrates how it is used. The
+probe emits the following information for each operation call:
+
+* AWS request ID
+* service ID
+* operation name
+* operation invocation time
+* http call time
+
+Each emitted message is prefixed with a unique (to a single operation)
+"invocation ID" which is set at the beginning of the handler stack.
+
+The entry point for instrumentation is exposed as `WithOperationTiming`, which
+is parameterized to accept a message handling function which will receive
+instrumentation "events" as formatted strings. `PrintfMSGHandler` is provided
+as a convenience which will simply dump messages to stdout.
+
+The service used here is interchangeable - ALL service client options accept
+`APIOptions` and an `HTTPClient` as configuration. For example,
+`WithOperationTiming` could instead be declared as:
+
+```go
+func WithOperationTiming(msgHandler func(string)) func(*s3.Options)
+func WithOperationTiming(msgHandler func(string)) func(*dynamodb.Options)
+// etc.
+```
+
+If you change it, be sure to change the signature of the function it returns as
+well.
+
+```
+import (
+    "context"
+    "fmt"
+    "log"
+    "net/http"
+    "sync"
+    "time"
+
+    awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+    awshttp "github.com/aws/aws-sdk-go-v2/aws/transport/http"
+    "github.com/aws/aws-sdk-go-v2/config"
+    "github.com/aws/aws-sdk-go-v2/service/sqs"
+    "github.com/aws/smithy-go/middleware"
+    smithyrand "github.com/aws/smithy-go/rand"
+)
+
+// WithOperationTiming instruments an SQS client to dump timing information for
+// the following spans:
+//   - overall operation time
+//   - HTTPClient call time
+//
+// This instrumentation will also emit the request ID, service name, and
+// operation name for each invocation.
+//
+// Accepts a message "handler" which is invoked with formatted messages to be
+// handled externally, you can use the declared PrintfMSGHandler to simply dump
+// these values to stdout.
+func WithOperationTiming(msgHandler func(string)) func(*sqs.Options) {
+    return func(o *sqs.Options) {
+        o.APIOptions = append(o.APIOptions, addTimingMiddlewares(msgHandler))
+        o.HTTPClient = &timedHTTPClient{
+            client:     awshttp.NewBuildableClient(),
+            msgHandler: msgHandler,
+        }
+    }
+}
+
+// PrintfMSGHandler writes messages to stdout.
+func PrintfMSGHandler(msg string) {
+    fmt.Printf("%s\n", msg)
+}
+
+type invokeIDKey struct{}
+
+func setInvokeID(ctx context.Context, id string) context.Context {
+    return middleware.WithStackValue(ctx, invokeIDKey{}, id)
+}
+
+func getInvokeID(ctx context.Context) string {
+    id, _ := middleware.GetStackValue(ctx, invokeIDKey{}).(string)
+    return id
+}
+
+func timeSpan(ctx context.Context, name string, consumer func(string)) func() {
+    start := time.Now()
+    return func() {
```

Comment:
for my own curiosity, whats the advantage of using a higher-order function here? when it looks like all use cases in this code sample of where `timeSpan` is called, it immediately calls the returned function

Reply:
It's not immediately called because of [defer](https://go.dev/tour/flowcontrol/12), I've clarified how it's used in a doc comment.

Reply:
i meant the second set of parentheses (i.e. the second invocation). the second invocation is called immediately after the first invocation.